### PR TITLE
Ignore special files when scanning.

### LINKED ISSUE #515: a non-regular file inside a filesystem directory magic-folder is pointed at prevents contents from being uploaded
If magic-folder is added pointed at a directory which itself contains something that is not a "regular" file - eg a directory or a symlink - then the contents of the directory are not uploaded.  The following appears repeatedly in the log:

```
af63034d-2bce-43e6-8bed-07748160ebbd
└── scanner:find-updates/1 ⇒ started 2021-07-28 19:13:08Z ⧖ 0.002s
    └── scanner:find-updates/2 ⇒ failed 2021-07-28 19:13:08Z
        ├── exception: exceptions.UnboundLocalError
        └── reason: local variable 'path_state' referenced before assignment

05a802a8-75d0-48d4-b9e2-c3d96c369d0a
└── eliot:traceback/1 2021-07-28 19:13:08Z
    ├── exception: exceptions.UnboundLocalError
    ├── reason: local variable 'path_state' referenced before assignment
    └── traceback: Traceback: <type 'exceptions.UnboundLocalError'>: local variable 'path_state' referenced before assignment⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/defer.py:501:errback⏎                                                                                                                                      
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/defer.py:568:_startRunCallbacks⏎                                                                                                                           
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/defer.py:654:_runCallbacks⏎                                                                                                                                
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/defer.py:1475:gotResult⏎                                                                                                                                   
        --- <exception caught here> ---⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/defer.py:1416:_inlineCallbacks⏎                                                                                                                            
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/python/failure.py:491:throwExceptionIntoGenerator⏎                                                                                                                  
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/eliot/_generators.py:141:wrapper⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/magic_folder/scanner.py:112:_scan⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/eliot/_generators.py:160:wrapper⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/twisted/internet/task.py:517:_oneWorkUnit⏎                                                                                                                                  
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/magic_folder/scanner.py:142:_process⏎
        /nix/store/j0q2p45gpj4w8p6zaixa3574a8c4lip4-python-2.7.17-env/lib/python2.7/site-packages/magic_folder/util/file.py:70:get_pathinfo⏎                                                                                                                                  
```

The expected behavior is:
* recursive handling will be applied to directories
* other supported filetypes are uploaded in a way that allows the downloader to recreate them
* unsupported filetypes at least do not get in the way of processing of all supported filetypes

### PR DESCRIPTION
**Pull request by tomprince**
Fixes #515.

### CODE REVIEW

**Review by meejah — APPROVED** [review]
overall looks good, some inline fixups/comments

File: src/magic_folder/scanner.py [hunk 1]
```diff
@@ -109,44 +110,66 @@ def process(path):
...
+                    if snapshot_state is not None:
+                        status.error_occurred(
+                            "File {} was a file, and now is {}.".format(
+                                relpath, "a directory" if path_info.is_dir else "not"
+                            )
+                        )
+                    return
+                if path_info.state != snapshot_state:
+                    # TODO: We may also want to compare checksums here,
+                    # to avoid `touch(1)` creating a new snapshot.
```
[the remark this answers is not in the document]
Comment (tomprince) [reply src/magic_folder/scanner.py:166]: I agree about the ticket, but this is an existing comment.

File: src/magic_folder/test/test_scanner.py [hunk 2]
```diff
@@ -168,11 +206,47 @@ def test_scan_something_local(self, name):
...
+        files = []
+        self.assertThat(
+            find_updated_files(
+                self.cooperator, self.config, files.append, status=self.folder_status
+            ),
+            succeeded(Always()),
+        )
+        self.assertThat(files, Equals([]))
+        self.assertThat(
+            self.status_service._folders["default"]["errors"],
```
[the remark this answers is not in the document]
Comment (tomprince) [reply src/magic_folder/test/test_scanner.py:240]: I'd prefer not to parse json in the test.

[the remark this answers is not in the document]
Comment (meejah) [reply src/magic_folder/test/test_scanner.py:240]: i figured it's better to do that then depend on definitely-internal state

[the remark this answers is not in the document]
Comment (tomprince) [reply src/magic_folder/test/test_scanner.py:240]: I don't feel too strongly about it, but I don't think it is a big deal either way.

**Pull request merged**